java/HealthVault Mobile/HVSample/src/main/java/com/microsoft/hsg/android/hvsample/PersonalImageLoader.java

```diff
@@ -0,0 +1,169 @@
+import com.microsoft.hsg.android.simplexml.client.RequestCallback;
+import com.microsoft.hsg.android.simplexml.methods.getthings3.request.ThingRequestGroup2;
+import com.microsoft.hsg.android.simplexml.methods.getthings3.response.ThingResponseGroup2;
+import com.microsoft.hsg.android.simplexml.things.thing.Thing2;
+import com.microsoft.hsg.android.simplexml.things.types.personalimage.PersonalImage;
+import com.microsoft.hsg.android.simplexml.things.types.types.Record;
+
+public class PersonalImageLoader implements ComponentCallbacks2 {
+
+	private ImageLruCache mCache;
+	private HealthVaultClient mHVClient;
+	private Activity mContext;
+	
+	private ImageView imageView;
+	
+	public PersonalImageLoader(Activity context,
+			HealthVaultClient client) {
+		ActivityManager activityManager = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
+        int memSize = activityManager.getMemoryClass() * 1024 * 1024;
+
+		mCache = new ImageLruCache(memSize);
+		mHVClient = client;
+        this.mContext = context;
+	}
+
+	public void load(String id, ImageView imageView, int defaultResource) {
+		imageView.setImageResource(defaultResource);
+        Bitmap image = mCache.get(id);
+        if (image != null) {
+            imageView.setImageBitmap(image);
+        }
+        else {
+        	List<Record> records = HealthVaultApp.getInstance().getRecordList();
+        	Record record = null;
+        	
+        	for(Record rcd : records) {
+        		if(id == rcd.getId()) {
+        			record = rcd;
+        			break;
+        		}
+        	}
+
+			mHVClient.asyncRequest(getImageAsync(record), new PersonalImageCallback(id, imageView));
+        }
+	}
+	
+	private Callable<Bitmap> getImageAsync(final Record record) {
+		return new Callable<Bitmap>() {
+    		public Bitmap call() throws URISyntaxException, IOException {
+    			// check if it exist in file
+    			File cacheDir = mContext.getCacheDir();
+    			
+    			if(!cacheDir.exists()) {
+    				cacheDir.mkdirs();
+    			}
+    			
+    			File file = new File(cacheDir, "personalimage" + record.getId());
+    			
+    			if(file.exists()) {
+    				try(FileInputStream in = new FileInputStream(file)) {
+						return BitmapFactory.decodeStream(in);
+					}
+    			}
+    			
+    			// try to get from web
+    			ThingResponseGroup2 response = record.getThings(ThingRequestGroup2.thingTypeQuery(PersonalImage.ThingType));
+    			List<Thing2> things = response.getThing();
+    			
+    			if(things != null && !things.isEmpty()) {
+    				Thing2 thing = things.get(0);
+					PersonalImage image = (PersonalImage)thing.getData();
+					
+					FileOutputStream destination = new FileOutputStream(file);
+					image.download(record, destination);
+					destination.flush();
+					destination.close();
+					
+					FileInputStream is = new FileInputStream(file);
+					Bitmap bitmap = BitmapFactory.decodeStream(is);
+					is.close();
```

Comment:
Might need to try catch this close as well to be safe and capture the error case.